### PR TITLE
Turn `temperature` and `top_p` into tensors

### PR DESCRIPTION
**Pull request by Liyang90**
Turn `temperature` and `top_p` into tensors, so there would be no recompilation for different `(temperature, top_p)` combinations. Note that temperature=0 and temperature>0 would still have different graphs because the computation logics are different.

### CODE REVIEW

File: llama/generation.py [hunk 1]
```diff
@@ -133,7 +140,7 @@ def generate(
 def sample_top_p(probs, p):
     probs_sort, probs_idx = torch.sort(probs, dim=-1, descending=True)
     probs_sum = torch.cumsum(probs_sort, dim=-1)
-    mask = probs_sum - probs_sort > p
+    mask = (probs_sum - probs_sort) > p
```
Comment (alanwaketan) [inline llama/generation.py:143]: Will this change the output?

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: llama/generation.py [hunk 2]
```diff
@@ -71,6 +72,10 @@ def generate(
         tokens = tokens.to(device)
         input_text_mask = tokens != self.tokenizer.pad_id
 
+        temperature_tensor = torch.tensor(temperature).to(device)
```
Comment (JackCaoG) [inline llama/generation.py:75]: nit, can you leave a comment explaining why we need to convert them to tensors?

File: llama/generation.py [hunk 3]
```diff
@@ -20,11 +20,12 @@ def __init__(self, model: Transformer, tokenizer: Tokenizer):
...
     def _generate_one_token(self, tokens, input_tokens, input_text_mask, cur_pos_tensor, 
-                            input_pos_tensor, output_pos_tensor, cache_kvs, temperature, top_p):
+                            input_pos_tensor, output_pos_tensor, cache_kvs,
+                            temperature_tensor, top_p_tensor, with_temp):
         logits, cache_kvs = self.model(input_tokens, input_pos_tensor, output_pos_tensor, cache_kvs)
-        if temperature > 0:
-            probs = torch.softmax(logits / temperature, dim=-1)
-            next_token = sample_top_p(probs, top_p)
+        if with_temp:
+            probs = torch.softmax(logits / temperature_tensor, dim=-1)
```
Comment (AlexWertheim) [inline llama/generation.py:27]: Just a quick sanity check here: is division by a singleton tensor the same as dividing by its value? LGTM, as long as we don't need to have (say) `temperature_tensor.item()` here

Comment (Liyang90) [reply llama/generation.py:27]: Right, it is the expectation, Dynamo doesn't allow any `.item()` call wrapped.

**Review by AlexWertheim — APPROVED** [review]
LGTM.

File: llama/generation.py (hunk 1 above)
Comment (Liyang90) [reply llama/generation.py:143]: No, this is just for better readability

File: llama/generation.py (hunk 2 above)
Comment (Liyang90) [reply llama/generation.py:77]: Added

**Review by JackCaoG — APPROVED** [review]

**Review by miladm — COMMENTED** [review]
Awesome. Thanks @Liyang90

**Comment by miladm** [comment]
I suggest we move this change to the `blog` branch also.

**Pull request merged**